client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssShuffle.java

```diff
@@ -0,0 +1,220 @@
+    this.indexReadLimit = jobConf.getInt(RssMRConfig.RSS_INDEX_READ_LIMIT,
+      RssMRConfig.RSS_INDEX_READ_LIMIT_DEFAULT_VALUE);
+    this.readBufferSize = (int)UnitConverter.byteStringAsBytes(
+      jobConf.get(RssMRConfig.RSS_CLIENT_READ_BUFFER_SIZE,
+        RssMRConfig.RSS_CLIENT_READ_BUFFER_SIZE_DEFAULT_VALUE));
+   }
+
+  protected MergeManager<K, V> createMergeManager(
+    ShuffleConsumerPlugin.Context context) {
+    return new MergeManagerImpl<K, V>(reduceId, jobConf, context.getLocalFS(),
+      context.getLocalDirAllocator(), reporter, context.getCodec(),
+      context.getCombinerClass(), context.getCombineCollector(),
+      context.getSpilledRecordsCounter(),
+      context.getReduceCombineInputCounter(),
+      context.getMergedMapOutputsCounter(), this, context.getMergePhase(),
+      context.getMapOutputFile());
+  }
+
+  @Override
+  public RawKeyValueIterator run() throws IOException, InterruptedException {
+
+    // get map-completion events to generate RSS taskIDs
+    final RssEventFetcher<K,V> eventFetcher =
+      new RssEventFetcher<K,V>(reduceId, umbilical, scheduler, this,
+        MAX_EVENTS_TO_FETCH);
+    eventFetcher.start();
+    while (!scheduler.waitUntilPrepareDone(PROGRESS_FREQUENCY)) {
+      reporter.progress();
+      synchronized (this) {
+        if (throwable != null) {
+          throw new Shuffle.ShuffleError("error in shuffle in " + throwingThreadName,
+            throwable);
+        }
+      }
+    }
+    eventFetcher.shutDown();
+    Roaring64NavigableMap taskIdBitmap = scheduler.getRssTaskIdBitmap();
+
+    // get assigned RSS servers
+    Set<ShuffleServerInfo> serverInfoSet = MRRssUtils.getAssignedServers(jobConf,
+        reduceId.getTaskID().getId());
+    List<ShuffleServerInfo> serverInfoList = new ArrayList<>();
+    for(ShuffleServerInfo server: serverInfoSet) {
+      serverInfoList.add(server);
+    }
+
+    // just get blockIds from RSS servers
+    ShuffleWriteClient writeClient = MRRssUtils.createShuffleClient(jobConf);
+    Roaring64NavigableMap blockIdBitmap = writeClient.getShuffleResult(
+      clientType, serverInfoSet, appId, 0, reduceId.getTaskID().getId());
+    writeClient.close();
+    scheduler.resetBlockCount(blockIdBitmap.getLongCardinality());
+
+    // start shuffle fetcher
+    CreateShuffleReadClientRequest request = new CreateShuffleReadClientRequest(
+      appId, 0, reduceId.getTaskID().getId(), storageType, basePath, indexReadLimit, readBufferSize,
+      partitionNumPerRange, partitionNum, blockIdBitmap, taskIdBitmap, serverInfoList, jobConf);
+    ShuffleReadClient shuffleReadClient = ShuffleClientFactory.getInstance().createShuffleReadClient(request);
+    RssFetcher fetcher = new RssFetcher(jobConf, reduceId, scheduler, merger, reporter, metrics, this, shuffleReadClient);
+    fetcher.start();
+
+    // Wait for shuffle to complete successfully
+    while (!scheduler.waitUntilDone(PROGRESS_FREQUENCY)) {
```

Comment:
Don't use the process like MR, it's bad design. It's hard to understand. It's not necessary to use multi-thread.

Reply:
Ok

Reply:
The multi-threaded fashion is updated single-threaded fashion.